server/src/main/java/org/spine3/server/aggregate/Aggregate.java

```diff
@@ -390,7 +413,30 @@ void restore(Snapshot snapshot) {
             builder.mergeFrom(stateToRestore);
             initVersion(versionFromSnapshot);
         } else {
+            injectState(stateToRestore, versionFromSnapshot);
+        }
+    }
+
+    /**
+     * Sets the passed state and version.
+     *
+     * <p>The method circumvents the protection in the {@link #setState(Message, Version)
+     * setState()} method by creating a fake builder instance, which is cleared
+     * after the call.
+     *
+     * <p>This method has package-private access to be accessible by the
+     * {@code AggregateBuilder} test utility class from the {@code testutil} module.
+     */
+    @VisibleForTesting
+    void injectState(S stateToRestore, Version versionFromSnapshot) {
+        try {
+            @SuppressWarnings("unchecked")
+                // The cast is safe as we checked the type on the construction.
+            final B fakeBuilder = (B) getState().newBuilderForType();
+            this.builder = fakeBuilder;
             setState(stateToRestore, versionFromSnapshot);
```

Comment:
If we are making the builder mandatory, then we could use it as well as an additional protection level.
E.g. what if we either 
* accept the `Builder` instances instead of `S`;
* or create the `Builder` instances manually from the `stateToRestore` passed with `injectState` method;

and then call something like `validate` on this builder before setting the state. Thus we'll protect from passing an invalid state, such as ignored Spine-custom field bounds (`required`, `min`/`max` etc).